ui/ducks/send/send.js

```diff
@@ -0,0 +1,1325 @@
+      );
+      if (
+        state.amount.mode === AMOUNT_MODES.MAX &&
+        state.asset.type === ASSET_TYPES.NATIVE
+      ) {
+        slice.caseReducers.updateAmountToMax(state);
+      }
+      slice.caseReducers.validateGasField(state);
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    /**
+     * sets the amount mode to the provided value as long as it is one of the
+     * supported modes (MAX|INPUT)
+     */
+    updateAmountMode: (state, action) => {
+      if (Object.values(AMOUNT_MODES).includes(action.payload)) {
+        state.amount.mode = action.payload;
+      }
+    },
+    updateAsset: (state, action) => {
+      state.asset.type = action.payload.type;
+      state.asset.balance = action.payload.balance;
+      if (state.asset.type === ASSET_TYPES.TOKEN) {
+        state.asset.details = action.payload.details;
+      } else {
+        // clear the details object when sending native currency
+        state.asset.details = null;
+        if (state.recipient.error === CONTRACT_ADDRESS_ERROR) {
+          // Errors related to sending tokens to their own contract address
+          // are no longer valid when sending native currency.
+          state.recipient.error = null;
+        }
+
+        if (state.recipient.warning === KNOWN_RECIPIENT_ADDRESS_ERROR) {
+          // Warning related to sending tokens to a known contract address
+          // are no longer valid when sending native currency.
+          state.recipient.warning = null;
+        }
+      }
+      // if amount mode is MAX update amount to max of new asset, otherwise set
+      // to zero. This will revalidate the send amount field.
+      if (state.amount.mode === AMOUNT_MODES.MAX) {
+        slice.caseReducers.updateAmountToMax(state);
+      } else {
+        slice.caseReducers.updateSendAmount(state, { payload: '0x0' });
+      }
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      // update the draft transaction
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    updateRecipient: (state, action) => {
+      state.recipient.address = action.payload.address ?? '';
+      state.recipient.nickname = action.payload.nickname ?? '';
+
+      if (state.recipient.address === '') {
+        // If address is null we are clearing the recipient and must return
+        // to the ADD_RECIPIENT stage.
+        state.stage = SEND_STAGES.ADD_RECIPIENT;
+      } else {
+        // if and address is provided and an id exists on the draft transaction,
+        // we progress to the EDIT stage, otherwise we progress to the DRAFT
+        // stage. We also reset the search mode for recipient search.
+        state.stage =
+          state.draftTransaction.id === null
+            ? SEND_STAGES.DRAFT
+            : SEND_STAGES.EDIT;
+        state.recipient.mode = RECIPIENT_SEARCH_MODES.CONTACT_LIST;
```

Comment:
Outside of scope, but noting: We probably could use the contact_list to save local address names.